### PR TITLE
Merge forward and central into one method

### PR DESCRIPTION
**Pull request by BichengYing**
it should simplified the on-fly-perturb impl. I didn't run the code one real machine, please double check, thanks!

### CODE REVIEW

File: gradient_estimators/random_gradient_estimator.py [hunk 1]
```diff
@@ -89,75 +87,53 @@ def put_grad(self, grad: torch.Tensor) -> None:
...
-
-            dir_grad = (pert_plus_loss - pert_minus_loss) / (2 * self.mu)
+            if self.grad_estimate_method == "central":
+                self.perturb_model(pb_norm, alpha=-2 * self.mu)
+                pert_minus_loss = criterion(self.model_forward(batch_inputs), labels)
+                self.perturb_model(pb_norm, alpha=self.mu)  # Restore model
+            elif self.grad_estimate_method == "forward":
+                self.perturb_model(pb_norm, alpha=-self.mu)  # Restore model
+
+            dir_grad = (pert_plus_loss - pert_minus_loss) / self.mu
```
Comment (ZidongLiu) [inline gradient_estimators/random_gradient_estimator.py:136]: / self.mu / 2 for central
/ self.mu for forward

Comment (ZidongLiu) [inline gradient_estimators/random_gradient_estimator.py:136]:
```suggestion
    if self.grad_estimate_method == "central":
            dir_grad = (pert_plus_loss - pert_minus_loss) / ( 2 * self.mu)
    else:
            dir_grad = (pert_plus_loss - pert_minus_loss) / self.mu
```

**Review by ZidongLiu — CHANGES_REQUESTED** [review]
1. feel some work might also need to happen in cezo_fl/shared.py.
2. divide by 2 times mu for central method

File: gradient_estimators/random_gradient_estimator.py (hunk 1 above)
Comment (BichengYing) [reply gradient_estimators/random_gradient_estimator.py]: good catch!

**Comment by BichengYing** [comment]
> 1. feel some work might also need to happen in cezo_fl/shared.py.

I checked the code. I didn't find any because there is no place explicitly calling `_forward_method` or `_central_method`. Can you double check?

**Review by ZidongLiu — APPROVED** [review]
> > 1. feel some work might also need to happen in cezo_fl/shared.py.
> 
> I checked the code. I didn't find any because there is no place explicitly calling `_forward_method` or `_central_method`. Can you double check?

Thought you are implementing layer-wise perturbation. LGTM

**Comment by ZidongLiu** [comment]
for 1 clients/ 1 sample/ 1 perturb: 
image
central works
image
forward works 

for 3 client / 2 sample / 5 perturb:
image
central works

image
forawrd works and is faster than central

**Comment by BichengYing** [comment]
why the speed in first two figure got slower and slower??

**Pull request merged**